File: Intersection Point in Y Shaped Linked Lists.py
''' Given two singly linked lists, write a program to get the point where two linked lists intersect each other. 
If the linked list does not merge at any point, it should return -1. '''

#User function Template for python3
'''
    Function to return the value at point of intersection
    in two linked list, connected in y shaped form.
    
    Function Arguments: head1, head2 (heads of both the lists)
    
    Return Type: Node # driver code will print the Node->data
'''
'''
# Node Class
class Node:
    def __init__(self, data):   # data -> value stored in node
        self.data = data
        self.next = None

'''

#Function to find intersection point in Y shaped Linked Lists.
class Solution:
    def intersectPoint(self, head1, head2):
        if not head1 or not head2:
            return None
        ptr1=head1
        ptr2=head2
        while(ptr1!=ptr2):
            ptr1=ptr1.next if ptr1 else head2
            ptr2=ptr2.next if ptr2 else head1
        return ptr1
        # code here

File: Intersection Point in Y Shaped Linked Lists.java
/* Given two singly linked lists, write a program to get the point where two linked lists intersect each other. 
If the linked list does not merge at any point, it should return -1. */


/* 
Node of a linked list
class Node {
    int data;
    Node next;
    Node(int d) {
        data = d;
        next = null;
    }
}

Linked List class
class LinkedList {
    Node head; // head of list
}
*/

class Solution {
    Node intersectPoint(Node head1, Node head2) {
        // Write your code here
        // Base Condition
        if (head1 == null || head2 == null) {
            return null;
        }
        Node temp1 = head1;
        Node temp2 = head2;
        while (temp1 != temp2) {
            temp1 = temp1.next;
            temp2 = temp2.next;
            if (temp1 == temp2) {
                return temp1;
            }
            if (temp1 == null) {
                temp1 = head2;
            }
            if (temp2 == null) {
                temp2 = head1;
            }
        }
        return temp1;
    }
}
